feature: Support validation groups with `@Validated`

kevin-wise opened the pull request:
Enable specifying the validation groups to be used with `@Validated`. For example, `@Validated(groups = Group1.class)`. Should this be based on 3.5.x, since it's a new feature?

Also, this is my first time trying to update the documentation. Hopefully I got the syntax right.

kevin-wise committed:
Support validation groups with `@Validated`

graemerocher commented:
Yes please target 3.5.x

kevin-wise commented:
> Yes please target 3.5.x

ok. When will 3.4.x be merged into 3.5.x? This PR depends on the fix that was just merged to 3.4.x.  As soon as that merges I'll rebase.

graemerocher commented:
I have merged up. Thanks.

kevin-wise commented:
Rebased and base branch updated. Should be ready for review

kevin-wise commented:
@graemerocher could I get a review? Would like to get this into 3.5.x if possible

graemerocher commented on validation/src/main/java/io/micronaut/validation/Validated.java:
```diff
@@ -39,4 +39,10 @@
+  /**
+   * The validation groups that will be used for validation.
+   */
+  Class<?>[] groups() default {};
```
Add `@since 3.5.0` to this method

kevin-wise replied:
done

graemerocher commented on validation/src/main/java/io/micronaut/validation/ValidatingInterceptor.java:
```diff
@@ -167,4 +179,9 @@ private Object validateReturnExecutableValidator(MethodInvocationContext<Object,
     }
+
+    private Class<?>[] getValidationGroups(MethodInvocationContext<Object, Object> context) {
+      final AnnotationValue<Validated> annotation = context.getAnnotation(Validated.class);
+      return annotation == null ? new Class<?>[0] : annotation.classValues("groups");
```
Rewrite to `context.classeValues(Validated.class, "group")`

kevin-wise replied:
done

graemerocher reviewed (changes requested):
Sorry for the delay, was away at a conference. Looks good but for a couple of minor things.

kevin-wise commented:
updated as requested, and rebased. No problem on the delay, I understand. Just didn't know what to expect since I'm new to this project. I hope tagging you was ok.

graemerocher commented:
tagging me is good, sometimes I need reminders :)

kevin-wise commented:
sorry for the additional force pushes. Just fixing the format on the javadoc to match convention. All done now, promise :)

kevin-wise commented:
@graemerocher does this need another review? Or can it get merged

sdelamo reviewed (approved):
Thanks for the contribution
